[E0403] False positive generated for implicit future type

### Code

```Rust
pub async fn test() {
    let a = async { loop { } };
    let a = std::pin::pin!(a);
    
    use std::future::Future;
    use std::pin::Pin;
    use std::task::{Poll, Context};
    
    #[allow(non_camel_case_types)]
    struct Foo<'a, a> {
        a: Pin<&'a mut a>,
    }
    
    #[allow(non_camel_case_types)]
    impl<'a, a: Future<Output = ()>> Future for Foo<'a, a> {
        type Output = ();
        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
            self.a.as_mut().poll(cx)
        }
    }
    
    Foo { a }.await
}
```


### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0403]: the name `a` is already used for a generic parameter in this item's generic parameters
  --> src/lib.rs:11:20
   |
3  |     let a = async { loop { } };
   |         - first use of `a`
...
11 |     struct Foo<'a, a> {
   |                    ^ already used

error[E0403]: the name `a` is already used for a generic parameter in this item's generic parameters
  --> src/lib.rs:16:14
   |
3  |     let a = async { loop { } };
   |         - first use of `a`
...
16 |     impl<'a, a: Future<Output = ()>> Future for Foo<'a, a> {
   |              ^ already used

error[E0277]: `Foo<'_, _>` is not a future
  --> src/lib.rs:23:15
   |
23 |     Foo { a }.await
   |              -^^^^^
   |              ||
   |              |`Foo<'_, _>` is not a future
   |              help: remove the `.await`
   |
   = help: the trait `Future` is not implemented for `Foo<'_, _>`, which is required by `Foo<'_, _>: IntoFuture`
   = note: Foo<'_, _> must be a future or must implement `IntoFuture` to be awaited
   = note: required for `Foo<'_, _>` to implement `IntoFuture`

Some errors have detailed explanations: E0277, E0403.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `playground` (lib) due to 3 previous errors
```


### Desired output

```Shell
No error code raised
```


### Rationale and extra context

This code snippet uses neither of the two error code conditions presented in https://doc.rust-lang.org/error_codes/E0403.html. 
It seems that the variable name `a` for the assignment `let a = async { loop { test::a().await; } };` is used implicitly in the type name, which would cause the struct declaration to shadow the first declaration.


### Other cases

_No response_

### Rust Version

```Shell
Nightly channel
Build using the Nightly version: 1.82.0-nightly
(2024-08-18 6de928dce9472b864f4e)
```


### Anything else?

_No response_

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -2677,14 +2677,14 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             // We also can't shadow bindings from associated parent items.
             for ns in [ValueNS, TypeNS] {
                 for parent_rib in self.ribs[ns].iter().rev() {
-                    seen_bindings
-                        .extend(parent_rib.bindings.keys().map(|ident| (*ident, ident.span)));
-
                     // Break at mod level, to account for nested items which are
                     // allowed to shadow generic param names.
                     if matches!(parent_rib.kind, RibKind::Module(..)) {
                         break;
                     }
+
+                    seen_bindings
+                        .extend(parent_rib.bindings.keys().map(|ident| (*ident, ident.span)));
                 }
             }
 

--- a/tests/ui/resolve/local-shadows-inner-generic.rs
+++ b/tests/ui/resolve/local-shadows-inner-generic.rs
@@ -1,0 +1,8 @@
+//@ check-pass
+
+#![allow(non_camel_case_types)]
+
+pub fn main() {
+    let a = 1;
+    struct Foo<a> { field: a, };
+}

